### PR TITLE
Mpi/experimental split communicator

### PR DESCRIPTION
**Pull request by jcotela**
First attempt at creating and using custom communicators instead of `MPI_COMM_WORLD`. Please consider this as a proof of concept for the eventual interface.

FYI @RiccardoRossi @adityaghantasala 

Note: in the current state (as of 04.09.2019) it requires #5517.

### CODE REVIEW

**Comment by jcotela** [comment]
Open question: should we have a Python interface for this?

**Comment by philbucher** [comment]
> Open question: should we have a Python interface for this?

speaking from the CoSimulation side: yes please :)

**Comment by pooyan-dadvand** [comment]
> Open question: should we have a Python interface for this?

I would create a Python interface for it. It can be used i.e. in MPI Montecarlo algorithms.

**Comment by jcotela** [comment]
After playing a bit with creating and using custom communicators, I have come to the conclusion that we need to be a bit more strict with the lifetime management of the `MPIDataCommunicator` objects. In this PR, essentially each `MPIDataCommunicator` acts as a wrapper around a single `MPI_Comm` and, if it is not one of the predefined solvers, takes care of freeing it on destruction (by calling `MPI_Comm_free`). To do this safely, I see two alternatives:

1. `MPIDataCommunicator` objects are non-copyable. Creating a new `MPIDataCommunicator` implies holding it with a unique_ptr (held by the `ParallelEnvironment` singleton) and all MPI calls involving this communicator are made through a `const DataCommunicator&` to the same object. The object is kept alive as long as the `ParallelEnvironment` lives (so until end of execution) or until explicitly unregistered from the ParallelEnvironment. In the later case, existing references will be left dangling, so one should be careful.

2.  We allow copying `MPIDataCommunicator`, but copying implies duplication of the underlying `MPI_Comm` by calling `MPI_Comm_dup`. This is a collective operation and involves duplicating the internal data that MPI keeps for the communicator, so it is probably not cheap. However, this is the only way I see of taking responsibility for freeing the `MPI_Comm` without doing our own reference counting. Compared to the previous approach, it is easier to use and could allow the use of "ad-hoc" communicators without going through the central register, but users can easily and inadvertently create an overhead of calls to `MPI_Comm_dup` and `MPI_Comm_free` by using value semantics.

The present PR goes in the direction of 1, although it still has some rough edges, which need to be fixed if we follow this direction.

**Comment by RiccardoRossi** [comment]
+1 for option 1

**Comment by jcotela** [comment]
I would say this is ready for a full review now. I followed option 1 above (using unique_ptr holders and disallowing copy construction) although this is not strongly enforced in the C++ layer (one can still create ad-hoc communicators without registering them). I can do that, but I would prefer to wait for a subsequent PR, since it will add a bit of noise (needs a trivial modification of most existing MPIDataCommunicator tests).

The "official" way of creating communicators in this PR is using the `DataCommunicatorFactory` methods, all of which register a new `DataCommunicator` in `ParallelEnvironment` internally and return a `const` reference to the created object. See tests (on C++ and Python) for examples of use.

**Comment by pooyan-dadvand** [comment]
Seems that I arrived late for this discussion. What about considering our DataCommunicator as a handler for mpi_comm? Then we can hold a shared_ptr to the mpi_comm and copy the DataCommunicator as much as we want without having problem with dangling references?

**Comment by jcotela** [comment]
@pooyan-dadvand I believe that a solution based on `shared_ptr<DataCommunicator>` could work. In any case, we would still need a wrapper class between `MPI_Comm` and the `shared_ptr`, because `MPI_Comm` does not clean up when it goes out of scope (or at least is not guaranteed to, as far as I can tell) so someone has to call `MPI_Comm_free` on it.

The main argument I would have against a shared_ptr based solution is that all communicators should be freed before `MPI_Finalize` and I don't see how this can be ensured if we let people keep shared pointers to them.

**Comment by jcotela** [comment]
Hi, we are discussing this issue with @philbucher and  @adityaghantasala and we think the safest way forward is the following:
1. If the communicator "comes from the outside" (typically MPI_COMM_WORLD), duplicate it internally when creating a MPIDataCommunicator.
2. Do not allow copying of DataCommunicators (or creating two communicators from the same MPIComm) to avoid freeing it twice. For this reason, creation of new DataCommunicators should be handled through the factory functions.
3. Keep using unique pointers to ensure that all communicators are freed before MPI_Finalize

**Comment by pooyan-dadvand** [comment]
My idea was `shared_ptr<MPI_Comm, MpiCommFree()>` and using it as a member of our `DataCommunicator`.  This would work as intended, isn't it?

**Comment by jcotela** [comment]
@pooyan-dadvand only if the shared pointer goes out of scope before `MPI_Finalize()` is called, but I do not trust it to be the case given our problems with python garbage collection.

**Comment by jcotela** [comment]
@pooyan-dadvand on a second reading, I am not sure I understood correctly. Do you mean having the shared pointer in addition to keeping the DataCommunicator itself in a unique pointer? Or keeping the DataCommunicator by value and only using the shared pointer internally?

The first would work, although I think that the double layer of pointers would be a bit cumbersome. The second runs into the usual lifetime problems if the DataCommunicator is alive after MPI_Finalize.

**Comment by pooyan-dadvand** [comment]
Now I see better the problem, my idea was to solve the problem of shared MPI_comm letting the DataCommunicator to be copied. Of course the `shared_ptr<MPI_Comm, MpiCommFree()>`  would be internal to DataCommunicator. At least I don't see if we want to expose it. 

The problem of being used after `MPI_Finalized` is another beast which I think your solution fits well. 

My only question is who would hold the unique pointers and is in charge of `MPI_Finalized`?

**Comment by jcotela** [comment]
@pooyan-dadvand you'll love this...
- The unique pointers are held in the `ParallelEnvironment` singleton and freed in its destructor.
- `MPI_Finalize` is called by the `MPIEnvironment` singleton destructor.
-  Calling the `InitializeMPIParallelRun()` function from the KratosMPI module instantiates first the `MPIEnvironment` and then the `ParallelEnvironment` in order, ensuring that the corresponding `static` instances will be deallocated in reverse order.
- the Kratos `__init__()` function takes care of importing `KratosMPI` and calling `InitializeMPIParallelRun()` before doing anything else if MPI is detected.

I would love to simplify this, and in particular have a single singleton instead of two, but `ParallelEnvironment` lives in the kratos core, so it can not call MPI functions directly. Note that my initial implementation called `InitializeMPIParallelRun` automatically on loading KratosMPI.so, but we decided against it. The argument is that this allows users to import the KratosMPI module on combined (serial + MPI compatible) scripts without having import errors or silently switching to MPI mode.

**Comment by pooyan-dadvand** [comment]
Really lovely..... Thanks for the explanation....

Maybe I am missing some part but where is the other singleton, beside the `PrallelEnvironment`?

**Comment by jcotela** [comment]
@pooyan-dadvand here:
https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/mpi/mpi_environment.h
(not part of this PR)

**Comment by pooyan-dadvand** [comment]
I agree that ideally one singleton should manage the MPI and all objects depends on it. But now I am a bit lost... I was thinking that `ParallelEvironment` is also manages the Singleton of MPI. While `MPIEnvironment` is not even derived from `ParallelEnvironment`. (Another misunderstanding from my side) So would you please give me (again) the complete image? Maybe in a call?

**Comment by jcotela** [comment]
@pooyan-dadvand we can agree on a meeting, but just a very rough picture of the main classes involved:
- MPIEnvironment (KratosMPICore): Singleton, manages MPI initialization/finalization.
- ParallelEnvironment (KratosCore): Singleton, manages DataCommunicators.
- DataCommunicator (KratosCore, do-nothing base class) and MPIDataCommunicator (KratosMPICore, derived implementation, wraps MPI) define the communication interface.
- Kernel keeps the "master" flag of being on MPI or not (`Kernel.IsMPIDistributedRun()`)
- Core `__init__.py` does the MPI detection on module load, sets up the Kernel flag and instantiates MPIEnvironment if needed.

I am not aware of any C++ idiom/pattern that combines singletons and derivation (since singletons are instantiated through a static method, which cannot be virtual). If there was a clean way to do it, I would happily use it to combine the two singleton classes.

**Comment by pooyan-dadvand** [comment]
Thank you very much!! Now becomes clear! At least I think!! :stuck_out_tongue: 

About hierarchical singleton, I think is possible to take advantage of having the `MPIDataCommunicator` derived from `DataCommunicator`. If you derive the `MPIEnvironment` from `ParallelEnvironment`, you can store the `MPIDataCommunicator` in the base class singleton and only change the constructor/destructor to manage the MPI  initialization/finalization. So basically all MPI comms will be active until the destruction of  `MPIEnvironment` in which the singleton is also cleaned `DataCommunicator`s

How do you see this?

**Comment by jcotela** [comment]
@pooyan-dadvand I like the idea. I had originally thought of something similar, but I could not find a way to cleanly select between base and derived singleton initialization. The problem as I see it is that the singleton is created by calling a static method of the class (which can not be virtual, so we cannot rely on polymorphism to select the correct version) and which takes no arguments (because otherwise each call to `Singleton::GetInstance()` could potentially be requesting a different object). In short, this is why I was asking if you know any idiom or pattern for that.

The only alternative approach I can think of is to rely on `Kernel::IsDistributed()` (yet another singleton) to determine whether this is an MPI run or not. I do not like this idea though, since this means I must be able to construct an "MPI-aware" kernel before even initializing MPI (which will likely also imply having the kernel alive after finalizing MPI), but it is an option.

**Comment by philbucher** [comment]
ping @pooyan-dadvand this is the main blocker for finishing the MPI-capabilities of the CoSimApp

**Comment by adityaghantasala** [comment]
@philbucher  it just occurred to me 
Once this is PR is done and we want to use it in CoSim  we cannot use `GetDefaultCommunicator`  for #5863 as we discussed.

**Comment by philbucher** [comment]
> @philbucher it just occurred to me
> Once this is PR is done and we want to use it in CoSim we cannot use `GetDefaultCommunicator` for #5863 as we discussed.

yes we will have  to think how to do it

But there are many places that will need refactoring. Thats why it would be good to start with it at some point

**Comment by adityaghantasala** [comment]
Agreed .. but was just saying it as I was preparing to make the PR we discussed ..

**Review by pooyan-dadvand — APPROVED** [review]
Re approving after the merge

**Comment by philbucher** [comment]
@jcotela and I were thinking that I would before merging try it in the CoSimApp, since this is probably the most prominent usecase

However I will probably need some more time to do this, hence I think we can go ahead and merge it as is. Probably not many will use it in the near future, hence we could still change things if necessary

**Review by sunethwarna — APPROVED** [review]
This is working with my mpi case :) :+1:

File: kratos/mpi/python/add_mpi_utilities_to_python.cpp [hunk 1]
```diff
@@ -19,12 +19,29 @@
...
 #include "mpi/utilities/model_part_communicator_utilities.h"
 #include "mpi/utilities/parallel_fill_communicator.h"
+#include "mpi/utilities/data_communicator_factory.h"
 #include "mpi/utilities/gather_modelpart_utility.h"
 #include "mpi/utilities/mpi_normal_calculation_utilities.h"
 
 namespace Kratos {
 namespace Python {
 
+const DataCommunicator& CreateFromListOfRanks(
```
Comment (philbucher) [inline kratos/mpi/python/add_mpi_utilities_to_python.cpp:29]: hm I don't see why this is necessary, pybind can convert list to vector 
Will try it out in my testings

**Review by philbucher — COMMENTED** [review]
still looks good to me, only one minor thought

Will merge it now and try it soon

**Pull request merged**